testTop/pyTestsApp/gateway_tests/conftest.py

```diff
@@ -0,0 +1,993 @@
+"""
+CA Gateway test configuration.
+
+
+Environment variables used:
+
+    EPICS_BASE
+    EPICS_HOST_ARCH
+    GATEWAY_PVLIST
+    GATEWAY_ROOT
+    IOC_EPICS_BASE or EPICS_BASE
+    VERBOSE
+    VERBOSE_GATEWAY
+
+"""
+
+import contextlib
+import dataclasses
+import gc
+import logging
+import math
+import os
+import subprocess
+import tempfile
+import textwrap
+import threading
+import time
+from concurrent.futures import ProcessPoolExecutor
+from typing import (Any, Dict, Generator, List, Mapping, Optional, Protocol,
+                    Tuple)
+
+import epics
+import pytest
+
+from . import config
+
+logger = logging.getLogger(__name__)
+
+
+@contextlib.contextmanager
+def run_process(
+    cmd: List[str],
+    env: Dict[str, str],
+    verbose: bool = True,
+    interactive: bool = False,
+    startup_time: float = 0.5,
+    wait_for: Optional[bytes] = None,
+):
+    """
+    Run ``cmd`` and yield a subprocess.Popen instance.
+
+    Parameters
+    ----------
+    cmd :
+    """
+    verbose = True
+    logger.info("Running: %s (verbose=%s)", " ".join(cmd), verbose)
+
+    proc = subprocess.Popen(
+        cmd,
+        env=env,
+        stdin=subprocess.PIPE,
+        stdout=subprocess.PIPE,
+        stderr=subprocess.STDOUT,
+    )
+
+    stdout = None
+    event = threading.Event()
+    assert proc.stdin is not None
+
+    def read_stdout():
+        """Read standard output in a background thread."""
+        nonlocal stdout
+        lines = []
+        startup_lines = []
+        t0 = time.monotonic()
+
+        assert proc.stdout is not None
+        while True:
+            line = proc.stdout.readline()
+            if not line:
+                break
+
+            if event.is_set():
+                startup_lines.append(line)
+            else:
+                lines.append(line)
+
+            if wait_for is not None and wait_for in line:
+                # Set the starting event if we see what we're waiting for
+                # to indicate the process is ready.
+                event.set()
+
+        stdout = b"".join(lines)
+        if verbose:
+            logger.warning(
+                "Read %d bytes from %s in %.1f sec",
+                len(stdout),
+                cmd[0],
+                time.monotonic() - t0,
+            )
+            if stdout:
+                try:
+                    stdout = stdout.decode("latin-1")
+                except Exception:
+                    stdout = str(stdout)
+
+                logger.warning(
+                    "Standard output for %s:\n" "    %s\n\n",
+                    cmd[0],
+                    textwrap.indent(str(stdout), "    "),
+                )
+
+    threading.Thread(daemon=True, target=read_stdout).start()
```

Comment:
This thread is never joined.  Since the function waits on the child misbehavior on interpreter exit is unlikely.  Still, this could instead be:

```py
slirp = threading.Thread(target=read_stdout)
slirp.start()

... yield proc
... proc.wait()

slirp.join()
```